Prohibit PHP closing tag at end of PHP-only file

jrfnl opened the pull request:
Not having a PHP closing tag if it would be the last thing in a file is best practice which should be adhered to as this can prevent PHP "headers already send" errors.

While this is not explicitly mentioned in the handbook, this can be considered to fall under the "Required: No PHP notices" rule.
Ref: https://make.wordpress.org/themes/handbook/review/required/explanations-and-examples/#code

grappler commented:
I am a bit on the fence about this. By having a PHP closing tag at the end of the file does not mean there will always be a PHP error.

I would like to see theme developers take responsibility for their code.

I would be ok seeing it merged and potentially removed again if there are number of complaints against it.

jrfnl commented:
@grappler The sniff as it is, throws an `error` with the message `A closing tag is not permitted at the end of a PHP file`.
We can change this to a `warning` and change the error message text to be milder via the ruleset.

Would that make you more comfortable about adding this sniff ? If so, any suggestions about the preferred error message ?

jrfnl commented:
Oh and this may be improtant to know as well: this sniff only checks files which *only* contain PHP code. So it doesn't check files with mixed PHP and HTML content.

joyously commented:
I'm for it, because I like preventing runtime errors.
Are you using the file extension for determining if it has only PHP? Or can the PHP output some HTML somewhere in it?

jrfnl commented:
> Are you using the file extension for determining if it has only PHP? Or can the PHP output some HTML somewhere in it?

No, the sniff actually checks the content of the file for inline HTML. PHP `echo`-ing out HTML is considered PHP for the purposes of this sniff.

grappler commented:
> Would that make you more comfortable about adding this sniff ?

Yes. It would be interesting to see if there are any complaints from theme authors when it is active.

> If so, any suggestions about the preferred error message ?

What about this message:
> To help prevent PHP "headers already send" errors PHP closing tag should be removed.

jrfnl commented:
@grappler OK, done. I've amended the existing commit as the change should be obvious.

* Turned the noticed into a warning.
* Changed the error message. I've used your suggestion as the basis and changed it to: `To help prevent PHP "headers already send" errors the PHP closing tag at the end of the file should be removed.`

justintadlock commented:
Again, this is something that should be discussed and not added without the approval of the TRT.  A warning-level notice (and not requirement/error) will probably be OK'd by the team.

This is only a potential problem for non-template files (headers already sent when templates get loaded) that have whitespace after the closing `?>`.  

And, there are some PHP-only templates in themes with closing `?>` tags.  Personally, I always recommend templates have distinct opening and closing PHP tags for all PHP because end users are far more likely to be copying/pasting code they don't understand in those.  A missing `?>` leaves room for more issues in my experience of running support.
